File: Sort/MergeSort/Command.cs
//
// Copyright (C) 2017, Christopher N. Hume.  All rights reserved.
//
// You should have received a copy of the MIT License along with this program.
// If not, see https://opensource.org/licenses/MIT.
//
// 2017-10-30 CNHume  Added Command class
//
namespace Sort {
  using System;

  public class Command {
    #region Virtual Fields
    public Int32? Length;
    public Int32? Merges;
    public Int32? InsertionLimit;
    public Boolean Print;
    #endregion

    #region Methods
    public void Parse(String[] args) {
      Length = null;
      InsertionLimit = null;
      Merges = null;
      Print = false;

      var usage = false;
      var count = args.Length;
      var n = 0;
      for (; n < count && !usage; n++) {
        var token = args[n];            // parse any switches
        var len = token.Length;
        if (len < 1 || token[0] != '-')
          break;                        // end of switches
        else if (len < 2)
          usage = true;
        else {                          // parse single character switch
          switch (token[1]) {
          case 'i':                     // the insertion-limit switch
            if (len > 2)                // whitespace optional
              usage = !TryParse(token.Substring(2, len - 2), out InsertionLimit);
            else if (n < count)         // whitespace allowed
              usage = !TryParse(args[++n], out InsertionLimit);
            else
              usage = true;
            break;

          case 'm':                     // the merges switch
            if (len > 2)                // whitespace optional
              usage = !TryParse(token.Substring(2, len - 2), out Merges);
            else if (n < count)         // whitespace allowed
              usage = !TryParse(args[++n], out Merges);
            else
              usage = true;
            break;

          case 'p':                     // the print switch
            if (len > 2)                // superfluous value specified
              usage = true;
            else
              Print = true;
            break;

          default:                      // switch unknown
            usage = true;
            break;
          }
        }
      }

      // length is required
      usage |= count <= n || !TryParse(args[n++], out Length);

      usage |= n < count;               // superfluous argument specified

      if (usage)                        // throw usage line if parse failed
        throw new ApplicationException("Usage: MergeSort [-i <insertion-limit>] [-m <merges>] [-p] length");
    }

    private static Boolean TryParse(String s, out Int32? value) {
      var valid = Int32.TryParse(s, out Int32 result);
      value = valid ? (Int32?)result : null;
      return valid;
    }
    #endregion
  }
}

File: C++/together/Command.cpp
// Copyright (C) 2018, Christopher N. Hume.  All rights reserved.
//
// 2018-05-16 CNHume  Added Command class
//
#include "Command.h"

#include <stdexcept>

string Command::arg_msg(const string& s, const string& name) {
  auto msg = string("Invalid ");
  msg += name;
  msg += ": ";
  msg += s;
  return msg;
}

uint32_t Command::parse_uint32(const char* s, const string& name) {
  char* pEnd;
  auto result = strtoul(s, &pEnd, 10);
  if (errno != 0 || *pEnd != '\0')
    throw invalid_argument(arg_msg(s, name));
  return result;
}

//
// command line parser
//
void Command::Parse(int argc, char* argv[]) {
  auto usage = false;
  auto n = 1;                           // skip verb
  for (; n < argc && !usage; n++) {     // parse any switches
    char* token = argv[n];
    auto len = strlen(token);
    if (len < 1 || token[0] != '-')     // end of switches
      break;
    else if (len < 2)
      usage = true;
    else {                              // parse single character switch
      switch (token[1]) {
      case 'd':                         // the display switch
        if (len > 2)                    // whitespace optional
          display = parse_uint32(&token[2], "display");
        else if (n < argc)              // whitespace allowed
          display = parse_uint32(argv[++n], "display");
        else
          usage = true;
        break;

      case 't':                         // the test switch
        if (len > 2)                    // superfluous value specified
          usage = true;
        else
          test = true;
        break;

      default:                          // switch unknown
        usage = true;
        break;
      }
    }
  }

  if (n < argc)                         // parse the filename
    filename = argv[n++];
  else                                  // filename is required
    usage = true;

  if (n < argc)                         // superfluous argument specified
    usage = true;

  if (usage)                            // throw usage line if parse failed
    throw runtime_error("Usage: together [-t] [-d <display-limit>] filename");
}
